DR2-1760 Add new indexed fields.

MancunianSam opened the pull request:
This looks like a lot but there's some bits I should have done initially
that I've tidied up here.

* The database builder will now write null, mapped to `Option` if a value
  is missing instead of empty string
* The webapp now has a get endpoint for search results with query
  parameters. The form POST just redirects to that.
* It now deletes and recreates the rows each time. This will allow us to
  handle updates.
* I've added the new fields which was the original intention.
* I've experimented a bit with property based testing with Scalacheck
  where we generate case classes with random data and see if they all
  pass.
* Some of the logging config was broken/annoying.

techncl commented on disaster-recovery-webapp/src/main/twirl/search.scala.html:
```diff
@@ -4,19 +4,11 @@
+            @input("zref", "ZRef")
+            @input("sourceId", "Source ID")
+            @input("citation", "Neutral Citation")
+            @date("Ingest date", "ingest-date")
```
You have a lowercased "d" here, whereas, it is uppercase elsewhere

techncl commented on disaster-recovery-webapp/src/test/scala/uk/gov/nationalarchives/webapp/FrontEndSpec.scala:
```diff
@@ -47,23 +50,110 @@ class FrontEndSpec extends AnyFlatSpec:
 
-  "OcflRoutes /search" should "render no results if no results are found" in {
+  "OcflRoutes /search GET" should "pass the correct parameters to the database call" in {
+    val id = UUID.randomUUID.toString
```
This doesn't seem to be used

MancunianSam replied:
Good spot, I've removed it.

techncl commented on disaster-recovery-webapp/src/main/scala/uk/gov/nationalarchives/webapp/FrontEndRoutes.scala:
```diff
@@ -1,33 +1,76 @@
+  private object IngestDateQueryParam extends OptionalQueryParamDecoderMatcher[Instant]("ingestDate")
+
+  case class SearchResponse(id: Option[UUID], zref: Option[String], sourceId: Option[String], citation: Option[String], ingestDateTime: Option[Instant]) {
+    private val params: Vector[(String, Option[String])] = Vector.from {
```
Do you need `Vector.from {}`? Couldn't you just do

```
      Vector(
        ("id", id.map(_.toString)),
        ("zref", zref),
        ("sourceId", sourceId),
        ("citation", citation),
        ("ingestDate", ingestDateTime.map(_.toEpochMilli.toString))
      ).filter(_._2.nonEmpty)
```

MancunianSam replied:
Yes you can, thanks. I've changed it.

techncl reviewed (commented):
Looks good! Just a few comments, as well as these

1. Do you think it's worth adding in some validation like:

  a. Date can't be after current date
  b. Month can't be over 13
  c. Day can't be larger than 32 etc?
2. for the test `"write" should "should write the values to the database"`, do you think it is worth checking if the DB is empty beforehand just incase the `createTable` in TestUtils is modified in the future (this test would catch it if it was)
3. I know you didn't add it in this PR, but is there a more descriptive name than `OCFL.generate`? Maybe `OCFL.generateOcflFileObjects`, `OCFL.generateOcflObjects` or`OCFL.generateFileObjects`?
4. In the OCFL spec, should there be a test for what would happen if the `Id` doesn't exist or if there is an error?
5. Should we wrap `strip()` around the values used to search? If so should we add a test for it?

MancunianSam commented:
1. For the validation:
a. Date can't be after current date. This feels a bit superfluous. If they try it, they'll get zero results. This isn't a public facing website so I think we'll be ok.
b. Month can't be over 13. I've added in the html attribute for min 1 max 12
c. Day can't be larger than 32 etc? Same here, I've used the html attributes.

2. For the test. Yeah, checking it's empty first is worth it in case there's an old database hanging around as well.
3. `generateOcflObjects` is ok. I've changed it.
4. If this Id doesn't exist, that means it's not in the OCFL repo which would be bad. I've added a test in for it.
5. I'm going to leave it. If it turns out that people are putting spaces where they didn't mean to then we can add it in later but I'd like to keep the validation as light as possible as I don't know how it'll be used.

I've also added some better error handling so if there's a problem you get a "There is a problem" page rather than the browsers default for server error.

MancunianSam commented on disaster-recovery-builder/src/main/scala/uk/gov/nationalarchives/builder/Database.scala:
```diff
@@ -30,11 +27,13 @@ object Database:
+        "insert into files (version, id, name, fileId, zref, path, fileName, ingestDateTime, sourceId, citation) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
+      val deleteAndInsert = for {
+        _ <- Update[UUID](deleteSql).updateMany(files.map(_.id))
+        _ <- Update[OcflFile](insertSql).updateMany(files)
```
I initially took it out because the return values in this for comprehension are `ConnectionIO[Int]` which can't be flatMapped with the `F[Unit]` that comes out of the logging calls. 

I've changed it a bit now so it returns `ConnectionIO[(Int, Int)]` with the delete and update count so we can have the message `s"$updateCount rows updated. $deleteCount rows deleted"`

MancunianSam commented on disaster-recovery-builder/src/main/scala/uk/gov/nationalarchives/builder/Database.scala:
```diff
@@ -19,8 +18,6 @@ trait Database[F[_]]:
 object Database:
   def apply[F[_]](using ev: Database[F]): Database[F] = ev
```
It's a really common thing in Scala tutorials, I think it is environment or something like it. I've changed it to `db`

MancunianSam commented on disaster-recovery-builder/src/main/scala/uk/gov/nationalarchives/builder/Main.scala:
```diff
@@ -44,7 +44,10 @@ object Main extends IOApp {
       .eval {
         for {
+          logger <- Slf4jLogger.create[IO]
+          _ <- logger.info("Receiving messages")
```
There is a similar message below. This was for debugging, I don't want to clutter up the logs with this every 20 seconds so I've taken it out.

I've added the `s"Received ${messages.length} messages from queue $queueUrl"` logging call inside the `IO.whenA` so we only log messages if we get any.

There's no real value to constantly outputting the same message, if the service fails for any reason, we'll get alerted when the queue starts filling up or the DLQ starts to get messages.

MancunianSam commented on disaster-recovery-builder/src/main/scala/uk/gov/nationalarchives/builder/Ocfl.scala:
```diff
@@ -45,27 +46,49 @@ object Ocfl:
+        .flatMap { metadata =>
+          {
+            val titleText = (metadata \ "InformationObject" \ "Title").text
+            if (titleText.isEmpty) None else Option(titleText)
```
I don't know. I don't want to throw errors for any missing data because there's always weirdness we don't account for. This is supposed to hold a copy of whatever is in Preservica. If there's no title then that's what we record. If one gets added later then we'll record that too.

MancunianSam commented on disaster-recovery-builder/src/test/scala/uk/gov/nationalarchives/builder/DatabaseSpec.scala:
```diff
@@ -30,6 +31,16 @@ class DatabaseSpec extends AnyFlatSpec with BeforeAndAfterEach:
     response.head should equal(file)
   }
 
+  "write" should "should write the updated values to the database if a row already exists" in {
```
Yep, all good points. I've changed the test to do that.

MancunianSam commented on disaster-recovery-webapp/src/main/twirl/date.scala.html:
```diff
@@ -0,0 +1,36 @@
+            </h1>
+        </legend>
+        <div class="govuk-date-input" id="dob">
+            <div class="govuk-date-input__item">
```
You've got the label and the id and the min and max values for validation. I think it'll make it a bit complicated so I'll leave it.

MancunianSam commented on disaster-recovery-webapp/src/main/twirl/date.scala.html:
```diff
@@ -0,0 +1,36 @@
+                    <label class="govuk-label govuk-date-input__label" for="@id-day">
+                        Day
+                    </label>
+                    <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="@id-day" name="@id-day" type="text" autocomplete="bday-day" inputmode="numeric">
```
Yep. I've changed it. The only reason GDS don't is so people can write "June" or "Sep" in the month field but I don't think we need to account for that.

techncl reviewed (approved):
👍
